### ISSUE TITLE
shellcheck does not recognize redeclaring variable from associate array to regular array.

### ISSUE THREAD

**Issue by mmlb** [koalaman/shellcheck#916]
#### For bugs
- Rule Id (if any, e.g. SC1000): SC2190
- My shellcheck version (`shellcheck --version` or "online"): online
- [x] I tried on shellcheck.net and verified that this is still a problem on the latest commit
- [ ] It's not reproducible on shellcheck.net, but I think that's because it's an OS, configuration or encoding issue

#### For new checks and feature suggestions
- [ ] shellcheck.net (i.e. the latest commit) currently gives no useful warnings about this
- [ ] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:

```sh

#!/bin/bash

declare -A indexes
indexes[$((RANDOM%2))]=true
indexes[$((RANDOM%2))]=true
indexes[$((RANDOM%2))]=true

# shellcheck disable=2207
indexes=($(echo "${!indexes[@]}" | tr ' ' '\n' | sort -n))

# now to show that shellcheck should be able to discover a new *regular* array being declared
indexes=()
unset indexes
declare -A indexes
indexes[$((RANDOM%2))]=true
indexes[$((RANDOM%2))]=true
indexes[$((RANDOM%2))]=true
# shellcheck disable=2116,2207
indexes=($(echo '[0]="hiworld"'))
# bash: indexes: $(echo '[0]="hiworld"'): must use subscript when assigning associative array
```

#### Here's what shellcheck currently says:
```
$ shellcheck myscript
 
Line 9:
indexes=($(echo "${!indexes[@]}" | tr ' ' '\n' | sort -n))
         ^-- SC2190: Elements in associative arrays need index, e.g. array=( [index]=value ) .
 
Line 18:
indexes=($(echo '[0]="hiworld"'))
         ^-- SC2190: Elements in associative arrays need index, e.g. array=( [index]=value ) .
```
#### Here's what I wanted or expected to see:

no error since I am redeclaring `indexes` as a normal array. shellcheck should be able to distinguish this due to the use of `$()` inside the declaration.

**Comment by mmlb** [comment]
actually it seems shellcheck just doesn't want to go from associative to indexed. Just ran the following on shellcheck.net:

```sh
#!/bin/bash

declare -A indexes
indexes[0]=true
echo ${#indexes[@]}

declare -a indexes
indexes=(0)
```
and the error is the same
```sh

$ shellcheck myscript
 
Line 8:
indexes=(0)
         ^-- SC2190: Elements in associative arrays need index, e.g. array=( [index]=value ) .

$ 
```

**Comment by nenadalm** [comment]
this happens also when the variables are completely separate in different functions (so they are local):
```
#!/bin/bash

function fnA() {
  declare -A indexes
  indexes[0]=true
  echo ${#indexes[@]}
}

function fnB() {
  declare -a indexes
  indexes=(0)
}
```

```
Line 11:
  indexes=(0)
           ^-- SC2190: Elements in associative arrays need index, e.g. array=( [index]=value ) .
```

**Comment by YetAnotherBugHunter** [comment]
I can confirm the behavior that @nenadalm demonstrates in his post. Changing the name of the variable made the false positives go away.

**Comment by GfEW** [comment]
This issue, even in its [cross function form](https://github.com/koalaman/shellcheck/issues/916#issuecomment-664143515), is still present in shellcheck 0.10.0.
